storage/redis/peer_store.go

```diff
@@ -0,0 +1,200 @@
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP)
+
+	return serializedPeer(b)
+}
+
+func panicIfClosed(closed <-chan struct{}) {
+	select {
+	case <-closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+}
+
+func ipType(ip net.IP) string {
+	if len(ip) == net.IPv6len {
+		return ipv6
+	}
+	return ipv4
+}
+
+func (s *peerStore) PutSeeder(infoHash bittorrent.InfoHash,
+	p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+
+	pk := newPeerKey(p)
+	return addPeer(s, infoHash, s.seederKeyPrefix+ipType(p.IP), pk)
+}
+
+func (s *peerStore) DeleteSeeder(infoHash bittorrent.InfoHash,
+	p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	pk := newPeerKey(p)
+	return removePeers(s, infoHash, s.seederKeyPrefix+ipType(p.IP), pk)
+}
+
+func (s *peerStore) PutLeecher(infoHash bittorrent.InfoHash,
+	p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	pk := newPeerKey(p)
+	return addPeer(s, infoHash, s.leecherKeyPrefix+ipType(p.IP), pk)
+
+}
+
+func (s *peerStore) DeleteLeecher(infoHash bittorrent.InfoHash,
+	p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	pk := newPeerKey(p)
+	return removePeers(s, infoHash, s.leecherKeyPrefix+ipType(p.IP), pk)
+
+}
+
+func (s *peerStore) GraduateLeecher(infoHash bittorrent.InfoHash,
+	p bittorrent.Peer) error {
+	panicIfClosed(s.closed)
+	err := s.PutSeeder(infoHash, p)
+	if err != nil {
+		return err
+	}
+	err = s.DeleteLeecher(infoHash, p)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Announce as many peers as possible based on the announcer being
+// a seeder or leecher
+func (s *peerStore) AnnouncePeers(infoHash bittorrent.InfoHash, seeder bool,
+	numWant int, announcer bittorrent.Peer) (peers []bittorrent.Peer,
+	err error) {
+	panicIfClosed(s.closed)
+	if numWant > s.maxNumWant {
+		numWant = s.maxNumWant
+	}
+
+	peers = []bittorrent.Peer{}
+	if seeder {
+		peers, err = getPeers(s, infoHash,
+			s.leecherKeyPrefix+ipType(announcer.IP),
+			numWant, peers, bittorrent.Peer{})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		peers, err = getPeers(s, infoHash,
+			s.seederKeyPrefix+ipType(announcer.IP),
+			numWant, peers, bittorrent.Peer{})
+		if err != nil {
+			return nil, err
+		}
+		if len(peers) < numWant {
+			peers, err = getPeers(s, infoHash,
+				s.leecherKeyPrefix+ipType(announcer.IP),
+				numWant, peers, announcer)
+		}
+	}
+	return peers, nil
+}
+
+func (s *peerStore) ScrapeSwarm(infoHash bittorrent.InfoHash, v6 bool) (
+	resp bittorrent.Scrape) {
+	panicIfClosed(s.closed)
+
+	ipType := ipv4
+	if v6 {
+		ipType = ipv6
+	}
+	complete, err := getPeersLength(s, infoHash,
+		s.seederKeyPrefix+ipType)
+	if err != nil {
+		return
+	}
+	resp.Complete = uint32(complete)
+	incomplete, err := getPeersLength(s, infoHash,
+		s.leecherKeyPrefix+ipType)
+	if err != nil {
+		return
+	}
+	resp.Incomplete = uint32(incomplete)
+	return
+}
+
+func (s *peerStore) Stop() <-chan error {
+	toReturn := make(chan error)
+	close(s.closed)
+	s.conn.Close()
```

Comment:
Is this something that could take a while? If yes, encapsulate this and `close(toReturn)` in a goroutine.